File: java/json-transform/src/main/java/ai/bizone/jsontransform/functions/TransformerFunctionPathJoin.java
package ai.bizone.jsontransform.functions;

import ai.bizone.jsontransform.functions.common.*;

import java.net.URLEncoder;
import java.nio.charset.StandardCharsets;
import java.util.Map;
import java.util.regex.Pattern;
import java.util.stream.Collectors;

public class TransformerFunctionPathJoin extends TransformerFunction {
    private static final String POSIX_SEPARATOR = "/";
    private static final String WINDOWS_SEPARATOR = "\\";
    private static final String WINDOWS_SEPARATOR_ESC = "\\\\";

    private static Pattern normalizePatternFactory(String s) {
        return Pattern.compile("([^." + s + "]+\\.|[^" + s + "]*[^." + s + "]|[^" + s + "]{3,})" + s + "\\.\\.($|" + s + ")");
    }
    private static final Pattern POSIX_DEDUPE_PATTERN = Pattern.compile(POSIX_SEPARATOR + "{2,}");
    private static final Pattern POSIX_DIRUP_PATTERN = normalizePatternFactory(POSIX_SEPARATOR);
    private static final Pattern POSIX_TRAILING_PATTERN = Pattern.compile(POSIX_SEPARATOR + "+$");
    private static final Pattern POSIX_SAMEDIR_PATTERN = Pattern.compile(POSIX_SEPARATOR + "([.]" + POSIX_SEPARATOR + ")+");
    private static final Pattern POSIX_SAMEDIR_START_PATTERN = Pattern.compile("^[.]" + POSIX_SEPARATOR);
    private static final Pattern WINDOWS_DEDUPE_PATTERN = Pattern.compile(WINDOWS_SEPARATOR_ESC + "{2,}");
    private static final Pattern WINDOWS_DIRUP_PATTERN = normalizePatternFactory(WINDOWS_SEPARATOR_ESC);
    private static final Pattern WINDOWS_TRAILING_PATTERN = Pattern.compile(WINDOWS_SEPARATOR_ESC + "+$");
    private static final Pattern WINDOWS_SAMEDIR_PATTERN = Pattern.compile(WINDOWS_SEPARATOR_ESC + "([.]" + WINDOWS_SEPARATOR_ESC + ")+");
    private static final Pattern WINDOWS_SAMEDIR_START_PATTERN = Pattern.compile("^[.]" + WINDOWS_SEPARATOR_ESC);

    public TransformerFunctionPathJoin() {
        super(FunctionDescription.of(
            Map.of(
            "type", ArgumentType.of(ArgType.String).position(0).defaultValue("POSIX")
            )
        ));
    }

    private static String normalizePath(String rawPath, String separator, boolean isURL) {
        var path = separator + rawPath;
        var isWin = separator.equals(WINDOWS_SEPARATOR);
        var escSep = isWin ? WINDOWS_SEPARATOR_ESC : separator; // need to escape for patterns replacement
        path = (isWin ? WINDOWS_DEDUPE_PATTERN : POSIX_DEDUPE_PATTERN).matcher(path).replaceAll(escSep); // dedupe all repeating separators

        // normalize dir-up
        var normalizer = isWin ? WINDOWS_DIRUP_PATTERN : POSIX_DIRUP_PATTERN;
        while (path.contains("..") && path.indexOf(separator + "..") != 0) {
            var lengthBefore = path.length();
            path = normalizer.matcher(path).replaceAll("");
            if (lengthBefore == path.length()) {
                break; // did nothing then break
            }
        }
        if (path.indexOf(separator + ".." + separator) == 0 || path.equals(separator + "..")) {
            throw new Error("Invalid path \"" + rawPath + "\"");
        }
        path = (isWin ? WINDOWS_TRAILING_PATTERN : POSIX_TRAILING_PATTERN).matcher(path).replaceAll(""); // remove trailing separators
        path = (isWin ? WINDOWS_SAMEDIR_PATTERN : POSIX_SAMEDIR_PATTERN).matcher(path).replaceAll(escSep);
        path = (isWin ? WINDOWS_SAMEDIR_START_PATTERN : POSIX_SAMEDIR_START_PATTERN).matcher(path).replaceAll(escSep);
        return isURL ? path : path.substring(1);
    }

    @Override
    public Object apply(FunctionContext context) {
        var arr = context.getJsonElementStreamer(null);
        if (arr == null) {
            return null;
        }
        var type = context.getEnum("type"); // should be either "/" or "\"
        var separator = type.equals("WINDOWS") ? WINDOWS_SEPARATOR : POSIX_SEPARATOR;

        var stream = arr.stream()
                .map(context::getAsString)
                .filter(x -> x != null && !x.isEmpty());
        var isURL = "URL".equals(type);
        if (isURL) {
            stream = stream.map(x -> URLEncoder.encode(x, StandardCharsets.UTF_8));
        }
        var path = stream.collect(Collectors.joining(separator));
        return normalizePath(path, separator, isURL);
    }
}

File: javascript/json-transform/src/functions/TransformerFunctionPathJoin.ts
import TransformerFunction from "./common/TransformerFunction";
import { ArgType } from "./common/ArgType";
import FunctionContext from "./common/FunctionContext";
import { getAsString, isNullOrUndefined } from "../JsonHelpers";
import { AsyncSequence } from "@wortise/sequency";

const POSIX_SEPARATOR = "/";
const WINDOWS_SEPARATOR = "\\";
const WINDOWS_SEPARATOR_ESC = "\\\\";

function NormalizePatternFactory(s: string) {
  return new RegExp("([^." + s + "]+\\.|[^" + s + "]*[^." + s + "]|[^" + s + "]{3,})" + s + "\\.\\.($|" + s + ")", "g");
}
const POSIX_DEDUPE_PATTERN = new RegExp(POSIX_SEPARATOR + "{2,}", "g");
const POSIX_DIRUP_PATTERN = NormalizePatternFactory(POSIX_SEPARATOR);
const POSIX_TRAILING_PATTERN = new RegExp(POSIX_SEPARATOR + "+$");
const POSIX_SAMEDIR_PATTERN = new RegExp(POSIX_SEPARATOR + "([.]" + POSIX_SEPARATOR + ")+", "g");
const POSIX_SAMEDIR_START_PATTERN = new RegExp("^[.]" + POSIX_SEPARATOR);
const WINDOWS_DEDUPE_PATTERN = new RegExp(WINDOWS_SEPARATOR_ESC + "{2,}", "g");
const WINDOWS_DIRUP_PATTERN = NormalizePatternFactory(WINDOWS_SEPARATOR_ESC);
const WINDOWS_TRAILING_PATTERN = new RegExp(WINDOWS_SEPARATOR_ESC + "+$");
const WINDOWS_SAMEDIR_PATTERN = new RegExp(WINDOWS_SEPARATOR_ESC + "([.]" + WINDOWS_SEPARATOR_ESC + ")+", "g");
const WINDOWS_SAMEDIR_START_PATTERN = new RegExp("^[.]" + WINDOWS_SEPARATOR_ESC);

class TransformerFunctionPathJoin extends TransformerFunction {
  constructor() {
    super({
      argsSet: [{ name: "type", type: ArgType.String, defaultValue: "POSIX" }],
    });
  }

  private static normalizePath(rawPath: string, separator: string, isURL: boolean): string {
    let path = separator + rawPath;
    const isWin = separator === WINDOWS_SEPARATOR;
    path = path.replace(isWin ? WINDOWS_DEDUPE_PATTERN : POSIX_DEDUPE_PATTERN, separator); // dedupe all repeating separators

    // normalize dir-up
    const normalizer = isWin ? WINDOWS_DIRUP_PATTERN : POSIX_DIRUP_PATTERN;
    while (path.includes("..") && path.indexOf(separator + "..") != 0) {
      const lengthBefore = path.length;
      path = path.replace(normalizer, "");
      if (lengthBefore == path.length) {
        break; // did nothing then break
      }
    }
    if (path.indexOf(separator + ".." + separator) == 0 || path === separator + "..") {
      throw new Error('Invalid path "' + rawPath + '"');
    }
    path = path.replace(isWin ? WINDOWS_TRAILING_PATTERN : POSIX_TRAILING_PATTERN, ""); // remove trailing separators
    path = path.replace(isWin ? WINDOWS_SAMEDIR_PATTERN : POSIX_SAMEDIR_PATTERN, separator);
    path = path.replace(isWin ? WINDOWS_SAMEDIR_START_PATTERN : POSIX_SAMEDIR_START_PATTERN, separator);
    return isURL ? path : path.substring(1);
  }

  override async apply(context: FunctionContext): Promise<any> {
    const arr = await context.getJsonElementStreamer(null);
    if (arr == null) {
      return null;
    }
    const type = await context.getEnum("type"); // should be either "/" or "\"
    const separator = type === "WINDOWS" ? WINDOWS_SEPARATOR : POSIX_SEPARATOR;

    let stream = (arr.stream().map(getAsString) as unknown as AsyncSequence<string>).filter(
      el => !isNullOrUndefined(el) && el !== "",
    );
    const isURL = type === "URL";
    if (isURL) {
      stream = stream.map(x => encodeURIComponent(x));
    }
    const path = await stream.joinToString({ separator });
    return TransformerFunctionPathJoin.normalizePath(path, separator, isURL);
  }
}

export default TransformerFunctionPathJoin;
